File: languages/js/src/entity/identity_v2/credential.ts
import { BloockBridge } from "../../bridge/bridge";
import {
  CredentialToJsonRequestV2,
  CredentialFromJsonRequestV2
} from "../../bridge/proto/identity_v2";
import * as identityEntitiesProto from "../../bridge/proto/identity_entities_v2";
import { NewConfigData } from "../../config/config";
import { CredentialProof } from "./credential_proof";
import { CredentialSchema } from "./credential_schema";
import { CredentialStatus } from "./credential_status";

export class Credential {
  context: string[];
  id: string;
  type: string[];
  issuanceDate: string;
  expiration: string;
  credentialSubject: any;
  credentialStatus: CredentialStatus;
  issuer: string;
  credentialSchema: CredentialSchema;
  proof: CredentialProof;

  constructor(
    context: string[],
    id: string,
    type: string[],
    issuanceDate: string,
    expiration: string,
    credentialSubject: any,
    credentialStatus: CredentialStatus,
    issuer: string,
    credentialSchema: CredentialSchema,
    proof: CredentialProof
  ) {
    this.context = context;
    this.id = id;
    this.type = type;
    this.issuanceDate = issuanceDate;
    this.expiration = expiration;
    this.credentialStatus = credentialStatus;
    this.credentialSubject = credentialSubject;
    this.issuer = issuer;
    this.credentialSchema = credentialSchema;
    this.proof = proof;
  }

  public toProto(): identityEntitiesProto.CredentialV2 {
    return identityEntitiesProto.CredentialV2.fromPartial({
      context: this.context,
      id: this.id,
      type: this.type,
      issuanceDate: this.issuanceDate,
      expiration: this.expiration,
      credentialStatus: this.credentialStatus.toProto(),
      credentialSubject: this.credentialSubject,
      issuer: this.issuer,
      credentialSchema: this.credentialSchema.toProto(),
      proof: this.proof.toProto()
    });
  }

  static fromProto(r: identityEntitiesProto.CredentialV2): Credential {
    return new Credential(
      r.context,
      r.id,
      r.type,
      r.issuanceDate,
      r.expiration,
      r.credentialSubject,
      CredentialStatus.fromProto(r.credentialStatus!),
      r.issuer,
      CredentialSchema.fromProto(r.credentialSchema!),
      CredentialProof.fromProto(r.proof!)
    );
  }

  public toJson(): Promise<string> {
    const bridge = new BloockBridge();

    const req = CredentialToJsonRequestV2.fromPartial({
      configData: NewConfigData(undefined),
      credential: this.toProto()
    });

    return bridge
      .getIdentityV2()
      .CredentialToJson(req)
      .then(res => {
        if (res.error) {
          throw res.error;
        }
        return res.json;
      });
  }

  static fromJson(json: string): Promise<Credential> {
    const bridge = new BloockBridge();

    const req = CredentialFromJsonRequestV2.fromPartial({
      configData: NewConfigData(undefined),
      json: json
    });

    return bridge
      .getIdentityV2()
      .CredentialFromJson(req)
      .then(res => {
        if (res.error) {
          throw res.error;
        }
        return Credential.fromProto(res.credential!);
      });
  }
}

File: languages/python/bloock/entity/identity_v2/credential.py
from __future__ import annotations

from typing import List

import bloock._bridge.proto.identity_entities_v2_pb2 as proto
from bloock._bridge import BloockBridge
from bloock._bridge.proto.identity_v2_pb2 import CredentialFromJsonRequestV2, CredentialToJsonRequestV2
from bloock._bridge.proto.shared_pb2 import Error
from bloock._config.config import Config
from bloock.entity.identity_v2.credential_proof import CredentialProof
from bloock.entity.identity_v2.credential_schema import CredentialSchema
from bloock.entity.identity_v2.credential_status import CredentialStatus


class Credential:
    def __init__(self, context: List[str], id: str, type: List[str], issuance_date: str, expiration: str, credential_subject: str,
                 credential_status: CredentialStatus, issuer: str, credential_schema: CredentialSchema,
                 proof: CredentialProof) -> None:
        self.context = context
        self.id = id
        self.type = type
        self.issuance_date = issuance_date
        self.expiration = expiration
        self.credential_subject = credential_subject
        self.credential_status = credential_status
        self.issuer = issuer
        self.credential_schema = credential_schema
        self.proof = proof

    @staticmethod
    def from_proto(c: proto.CredentialV2) -> Credential:
        return Credential(
            context=c.context,
            id=c.id,
            type=c.type,
            issuance_date=c.issuance_date,
            expiration=c.expiration,
            credential_subject=c.credential_subject,
            credential_status=CredentialStatus.from_proto(c.credential_status),
            issuer=c.issuer,
            credential_schema=CredentialSchema.from_proto(c.credential_schema),
            proof=CredentialProof.from_proto(c.proof)
        )

    def to_proto(self) -> proto.CredentialV2:
        return proto.CredentialV2(
            context=self.context,
            id=self.id,
            type=self.type,
            issuance_date=self.issuance_date,
            expiration=self.expiration,
            credential_subject=self.credential_subject,
            credential_status=self.credential_status.to_proto(),
            issuer=self.issuer,
            credential_schema=self.credential_schema.to_proto(),
            proof=self.proof.to_proto()
        )

    @staticmethod
    def from_json(json: str) -> proto.CredentialV2:
        bridge = BloockBridge()

        req = CredentialFromJsonRequestV2(
            config_data=Config.default(),
            json=json
        )

        res = bridge.identity_v2().CredentialFromJson(req)
        if res.error != Error():
            raise Exception(res.error.message)

        return Credential.from_proto(res.credential)

    def to_json(self) -> str:
        bridge = BloockBridge()

        req = CredentialToJsonRequestV2(
            config_data=Config.default(),
            credential=self.to_proto()
        )

        res = bridge.identity_v2().CredentialToJson(req)
        if res.error != Error():
            raise Exception(res.error.message)

        return res.json
